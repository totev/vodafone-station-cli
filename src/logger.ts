/* eslint-disable no-console */

export interface Log{
    log(message?: string, ...args: unknown[]): void;
    warn(input: string | Error | unknown): void;
    debug(...args: unknown[]): void;
    error(input: string | Error | unknown, ...options: unknown[]): void;
}

export class OclifLogger implements Log {
  // eslint-disable-next-line no-useless-constructor
  constructor(private delegateLog: Function, private delegateWarn: Function, private delegateDebug: Function, private delegateError: Function) {
  }

  error(input: string | Error, options: { code?: string | undefined; exit: false }): void {
    this.delegateError(input, options)
  }

  log(message?: string, ...args: unknown[]): void {
    this.delegateLog(message, args)
  }

  warn(input: string | Error): void {
    this.delegateWarn(input)
  }

  debug(...args: unknown[]): void {
    this.delegateDebug(args)
  }
}

export class ConsoleLogger implements Log {
  error(input: string | Error, options: { code?: string | undefined; exit: false }): void {
    console.log(input, options)
  }

  debug(...args: unknown[]): void {
    console.debug(args)
  }

  warn(input: string | Error): void {
    console.warn(input)
  }

  log(message?: string, ...args: unknown[]): void {
    console.log(message, args)
  }
}
