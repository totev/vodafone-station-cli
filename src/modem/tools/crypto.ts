/* eslint-disable new-cap */
import sjcl from 'sjcl'
const SJCL_ITERATIONS = 1000
const SJCL_KEYSIZEBITS = 128
const SJCL_TAGLENGTH = 128

export function prepareParameters(
  derivedKey: string,
  ivHex: string,
  authData: string
) {
  return {
    prf: new sjcl.cipher.aes(sjcl.codec.hex.toBits(derivedKey)),
    iv: sjcl.codec.hex.toBits(ivHex),
    authData: sjcl.codec.utf8String.toBits(authData),
  }
}

export function deriveKey(password: string, salt: string): string {
  const derivedKeyBits = sjcl.misc.pbkdf2(
    password,
    sjcl.codec.hex.toBits(salt),
    SJCL_ITERATIONS,
    SJCL_KEYSIZEBITS
  )

  return sjcl.codec.hex.fromBits(derivedKeyBits)
}

export function deriveKeyTechnicolor(password: string, salt: string): string {
  const derivedKeyBits = sjcl.misc.pbkdf2(password, salt,     SJCL_ITERATIONS,
    SJCL_KEYSIZEBITS)
  return sjcl.codec.hex.fromBits(derivedKeyBits)
}

export function encrypt(
  derivedKey: string,
  plainText: string,
  ivHex: string,
  authData: string
): string {
  const bitParams = prepareParameters(derivedKey, ivHex, authData)
  const encryptedBits = sjcl.mode.ccm.encrypt(
    bitParams.prf,
    sjcl.codec.utf8String.toBits(plainText),
    bitParams.iv,
    bitParams.authData,
    SJCL_TAGLENGTH
  )
  return sjcl.codec.hex.fromBits(encryptedBits)
}

export function decrypt(
  derivedKey: string,
  cipherTextHex: string,
  ivHex: string,
  authData: string
): string {
  const bitParams = prepareParameters(derivedKey, ivHex, authData)
  const decryptedBits = sjcl.mode.ccm.decrypt(
    bitParams.prf,
    sjcl.codec.hex.toBits(cipherTextHex),
    bitParams.iv,
    bitParams.authData,
    SJCL_TAGLENGTH
  )
  return sjcl.codec.utf8String.fromBits(decryptedBits)
}

