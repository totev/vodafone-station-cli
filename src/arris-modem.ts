import {Log} from './logger'
import {Modem} from './modem'

export class Arris extends Modem {
  constructor(readonly modemIp: string, readonly logger: Log) {
    super(modemIp, logger)
  }

  async logout(): Promise<void> {
    throw new Error('Not implemented!')
  }
}
