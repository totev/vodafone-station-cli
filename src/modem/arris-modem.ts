import {Log} from '../logger'
import {DocsisStatus, HumanizedDocsis31ChannelStatus, HumanizedDocsisChannelStatus, Modem, DocsisChannelType, ExposedHostSettings, HostExposureSettings, Protocol} from './modem'
import {decrypt, deriveKey, encrypt} from './tools/crypto'
import {CryptoVars, extractCredentialString, extractCryptoVars, extractDocsisStatus} from './tools/html-parser'

export interface ArrisDocsisStatus {
  downstream: ArrisDocsisChannelStatus[];
  upstream: ArrisDocsisChannelStatus[];
  downstreamChannels: number;
  upstreamChannels: number;
  ofdmChannels: number;
  time: string;
}

export interface ArrisDocsisChannelStatus {
  ChannelID: string;
  ChannelType: DocsisChannelType;
  Frequency: string|number;
  Modulation: string;
  PowerLevel: string;
  SNRLevel?: string | number;
  LockStatus: string;
}

export interface SetPasswordRequest {
  AuthData: string;
  EncryptData: string;
  Name: string;
}

export interface SetPasswordResponse {
  p_status: string;
  encryptData: string;
  p_waitTime?: number;
}

interface ArrisGetHostExposureSettings {
  hostExposure: ArrisGetExposedHostSettings[];
  dhcpclient: any;
}

interface ArrisGetExposedHostSettings {
  ServiceName: string;
  MAC: string;
  Protocol: Protocol;
  StartPort: number;
  EndPort: number;
  Status: string;
  Index: string;
}

interface ArrisSetHostExposureSettings {
  hEditRule: ArrisSetExposedHostSettings[];
}

interface ArrisSetExposedHostSettings {
  name: string;
  macAddress: string;
  protocol: Protocol;
  startPort: number;
  endPort: number;
  enable: string;
  index: string;
}

export function normalizeChannelStatus(channelStatus: ArrisDocsisChannelStatus): HumanizedDocsisChannelStatus | HumanizedDocsis31ChannelStatus {
  const frequency: Record<string, number> = {}
  if (channelStatus.ChannelType === 'SC-QAM') {
    frequency.frequency = channelStatus.Frequency as number
  }

  if (['OFDM', 'OFDMA'].includes(channelStatus.ChannelType)) {
    const ofdmaFrequency = String(channelStatus.Frequency).split('~')
    frequency.frequencyStart = Number(ofdmaFrequency[0])
    frequency.frequencyEnd = Number(ofdmaFrequency[1])
  }

  const powerLevel = parseFloat(channelStatus.PowerLevel.split("/")[0]);
  const snr = parseInt(`${channelStatus.SNRLevel ?? 0}`, 10);
  return {
    channelId: channelStatus.ChannelID,
    channelType: channelStatus.ChannelType,
    modulation: channelStatus.Modulation,
    powerLevel: isNaN(powerLevel) ? -100 : powerLevel,
    lockStatus: channelStatus.LockStatus,
    snr: isNaN(snr) ? 0 : snr,
    ...frequency,
  } as HumanizedDocsisChannelStatus | HumanizedDocsis31ChannelStatus;
}

export function normalizeDocsisStatus(arrisDocsisStatus: ArrisDocsisStatus): DocsisStatus {
  const result: DocsisStatus = {
    downstream: [],
    downstreamOfdm: [],
    upstream: [],
    upstreamOfdma: [],
    time: arrisDocsisStatus.time
  }
  result.downstream = arrisDocsisStatus.downstream
    .filter(downstream => downstream.ChannelType === 'SC-QAM')
    .map(normalizeChannelStatus) as HumanizedDocsisChannelStatus[]

  result.downstreamOfdm = arrisDocsisStatus.downstream
    .filter(downstream => downstream.ChannelType === 'OFDM')
    .map(normalizeChannelStatus) as HumanizedDocsis31ChannelStatus[]

  result.upstream = arrisDocsisStatus.upstream
    .filter(upstream => upstream.ChannelType === 'SC-QAM')
    .map(normalizeChannelStatus) as HumanizedDocsisChannelStatus[]

  result.upstreamOfdma = arrisDocsisStatus.upstream
    .filter(upstream => upstream.ChannelType === 'OFDMA')
    .map(normalizeChannelStatus) as HumanizedDocsis31ChannelStatus[]
  return result
}

export class Arris extends Modem {
  private csrfNonce = ''
  constructor(readonly modemIp: string, readonly logger: Log) {
    super(modemIp, logger)
  }

  async logout(): Promise<void> {
    try {
      this.logger.log('Logging out...')
      return  this.httpClient.post('/php/logout.php')
    } catch (error) {
      this.logger.error('Could not do a full session logout', error)
      throw error
    }
  }

  async login(password: string): Promise<void> {
    const cryptoVars = await this.getCurrentCryptoVars()
    const encPw = this.encryptPassword(password, cryptoVars)
    this.logger.debug('Encrypted password: ', encPw)
    const serverSetPassword = await this.createServerRecord(encPw)
    this.logger.debug('ServerSetPassword: ', serverSetPassword)

    const csrfNonce = this.loginPasswordCheck(
      serverSetPassword.encryptData,
      cryptoVars,
      deriveKey(password, cryptoVars.salt)
    )
    this.logger.debug('Csrf nonce: ', csrfNonce)

    await this.addCredentialToCookie()
    this.csrfNonce = csrfNonce
  }

  async  getCurrentCryptoVars(): Promise<CryptoVars> {
    try {
      const {data} = await this.httpClient.get('/', {
        headers: {Accept: 'text/html,application/xhtml+xml,application/xml'},
      })
      const cryptoVars = extractCryptoVars(data as string)
      this.logger.debug('Parsed crypto vars: ', cryptoVars)
      return cryptoVars
    } catch (error) {
      this.logger.error('Could not get the index page from the router', error)
      throw error
    }
  }

  encryptPassword(
    password: string,
    cryptoVars: CryptoVars
  ): SetPasswordRequest {
    const jsData =
    '{"Password": "' + password + '", "Nonce": "' + cryptoVars.sessionId + '"}'
    const key = deriveKey(password, cryptoVars.salt)
    const authData = 'loginPassword'
    const encryptData = encrypt(key, jsData, cryptoVars.iv, authData)

    return {
      EncryptData: encryptData,
      Name: Modem.USERNAME,
      AuthData: authData,
    }
  }

  loginPasswordCheck(
    encryptedData: string,
    cryptoVars: CryptoVars,
    key: string
  ): string {
    const csrfNonce = decrypt(key, encryptedData, cryptoVars.iv, 'nonce')
    return csrfNonce
  }

  async  createServerRecord(
    setPasswordRequest: SetPasswordRequest
  ): Promise<SetPasswordResponse> {
    try {
      const {data} = await this.httpClient.post<SetPasswordResponse>(
        '/php/ajaxSet_Password.php',
        setPasswordRequest
      )
      // TODO handle wrong password case
      // { p_status: 'Lockout', p_waitTime: 1 }
      if (data.p_status === 'Lockout') {
        throw new Error(`Remote user locked out for: ${data.p_waitTime}s`)
      }
      return data
    } catch (error) {
      this.logger.error('Could not pass password on remote router.', error)
      throw error
    }
  }

  async addCredentialToCookie(): Promise<void> {
    const credential = await this.fetchCredential()
    this.logger.debug('Credential: ', credential)
    // set obligatory static cookie
    this.cookieJar.setCookie(`credential= ${credential}`, `http://${this.modemIp}`)
  }

  async  fetchCredential(): Promise<string> {
    try {
      const {data} = await this.httpClient.get('/base_95x.js')
      return extractCredentialString(data as string)
    } catch (error) {
      this.logger.error('Could not fetch credential.', error)
      throw error
    }
  }

  async docsis(): Promise<DocsisStatus> {
    if (!this.csrfNonce) {
      throw new Error('A valid csrfNonce is required in order to query the modem.')
    }
    try {
      const {data} = await this.httpClient.get('/php/status_docsis_data.php', {
        headers: {
          csrfNonce: this.csrfNonce,
          Referer: `http://${this.modemIp}/?status_docsis&mid=StatusDocsis`,
          Connection: 'keep-alive',
        },
      })
      return normalizeDocsisStatus(extractDocsisStatus(data as string))
    } catch (error) {
      this.logger.error('Could not fetch remote docsis status', error)
      throw error
    }
  }

  async restart(): Promise<unknown> {
    try {
      const {data} = await this.httpClient.post(
        'php/ajaxSet_status_restart.php',
        {
          RestartReset: 'Restart',
        },
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?status_docsis&mid=StatusDocsis`,
            Connection: 'keep-alive',
          },
        }
      )
      this.logger.log('Router is restarting')
      return data
    } catch (error) {
      this.logger.error('Could not restart router.', error)
      throw error
    }
  }

  async firewall(status:boolean = true): Promise<void> {
    const action = status ? 'on' : 'off'
    try {
      const {data} = await this.httpClient.post(
        'api/v1/firewall',
        {
          FirewallLevel: action,
          FirewallLevelV6: action
        },
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?status_docsis&mid=StatusDocsis`,
            Connection: 'keep-alive',
          },
        }
      )
      this.logger.log('Firewall is enabled')
      return data
    } catch (error) {
      this.logger.error('Could not enable firewall.', error)
      throw error
    }
  }

  _convertGetExposedHostSettings(settings: ArrisGetExposedHostSettings): ExposedHostSettings {
    return {
      serviceName: settings.ServiceName,
      mac: settings.MAC,
      protocol: settings.Protocol,
      startPort: settings.StartPort,
      endPort: settings.EndPort,
      enabled: settings.Status === "Enabled" ? true : false,
      index: Number.parseInt(settings.Index),
    } as ExposedHostSettings
  }

  async getHostExposure(): Promise<HostExposureSettings> {
    try {
      const {data} = await this.httpClient.get(
        'php/net_ipv6_host_exposure_data.php?{"hostExposure":{},"dhcpclient":{}}',
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?net_ipv6_host_exposure&mid=NetIPv6HostExposure`,
            Connection: 'keep-alive',
          },
        }
      )
      return {
        hosts: (data as ArrisGetHostExposureSettings)
          .hostExposure.map(this._convertGetExposedHostSettings)
      } as HostExposureSettings
    }
    catch (error) {
      this.logger.error("Could not get host exposure data:\n", error)
      throw error
    }
  }

  _convertSetExposedHostSettings(settings: ExposedHostSettings): ArrisSetExposedHostSettings {
    return {
      name: settings.serviceName,
      macAddress: settings.mac,
      protocol: settings.protocol,
      startPort: settings.startPort,
      endPort: settings.endPort,
      enable: settings.enabled ? "Enabled" : "Disabled",
      index: settings.index.toString(),
    }
  }

  async setHostExposure(settings: HostExposureSettings): Promise<void> {
    const convertedSettings =
      {hEditRule: settings.hosts.map(this._convertSetExposedHostSettings)} as ArrisSetHostExposureSettings
    try {
      await this.httpClient.post(
        'php/ajaxSet_net_ipv6_host_exposure_data.php',
        convertedSettings,
        {
          headers: {
            csrfNonce: this.csrfNonce,
            Referer: `http://${this.modemIp}/?net_ipv6_host_exposure&mid=NetIPv6HostExposure`,
            Connection: 'keep-alive',
          }
        }
      )
    }
    catch (error) {
      console.error("Could not set host exposure data:\n", error)
      throw error
    }
  }
}

