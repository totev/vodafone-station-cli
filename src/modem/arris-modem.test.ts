import {ConsoleLogger} from '../logger'
import {Arris, ArrisDocsisStatus, normalizeDocsisStatus} from './arris-modem'
import {CryptoVars} from './tools/html-parser'
import fixtureDocsisStatus from './__fixtures__/docsisStatus_arris.json'
import fixtureDocsisStatusOfdma from './__fixtures__/docsisStatus_ofdm_arris.json'

describe('Arris', () => {
  test('should encrypt', () => {
    const expected = {
      EncryptData:
        '8059e124da83bf88c89ae02ab0c7a0335a272a2045e5c3fb075dcfba42788b5436483e01f37b5f25c50b5d9e6366734eb9eb33919d892e97bb025c63e35b5a76e5ffe53a292e8e8ebc99c923ea2977803b',
      Name: 'admin',
      AuthData: 'loginPassword',
    }
    const given: CryptoVars = {
      iv: 'c68af53914949158',
      salt: 'c4d0a0c70c3fcac4',
      sessionId: '01a91cedd129fd8c6f18e3a1b58d096f',
      nonce: 'WslSZgE7NuQr+1BMqiYEOBMzQlo=',
    }
    const arrisModem = new Arris('0.0.0.0', new ConsoleLogger())
    expect(arrisModem.encryptPassword('test', given)).toEqual(expected)
  })

  describe('normalizeDocsisStatus', () => {
    test('should work with ofdm in download', () => {
      const status = normalizeDocsisStatus(fixtureDocsisStatus as ArrisDocsisStatus)
      expect(status).toMatchSnapshot()
    })
    test('should work with ofdm in download and in upload', () => {
      const status = normalizeDocsisStatus(fixtureDocsisStatusOfdma as ArrisDocsisStatus)
      expect(status).toMatchSnapshot()
    })
  })
})
