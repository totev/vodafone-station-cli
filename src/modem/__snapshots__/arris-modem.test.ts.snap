// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arris normalizeDocsisStatus should work with ofdm in download 1`] = `
{
  "downstream": [
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 130,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -6.7,
      "snr": 35,
    },
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 138,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -6.4,
      "snr": 35,
    },
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 146,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -6.7,
      "snr": 35,
    },
    {
      "channelId": "5",
      "channelType": "SC-QAM",
      "frequency": 602,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.7,
      "snr": 38,
    },
    {
      "channelId": "6",
      "channelType": "SC-QAM",
      "frequency": 618,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.8,
      "snr": 38,
    },
    {
      "channelId": "7",
      "channelType": "SC-QAM",
      "frequency": 626,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.6,
      "snr": 38,
    },
    {
      "channelId": "8",
      "channelType": "SC-QAM",
      "frequency": 642,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.2,
      "snr": 38,
    },
    {
      "channelId": "9",
      "channelType": "SC-QAM",
      "frequency": 650,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.8,
      "snr": 38,
    },
    {
      "channelId": "10",
      "channelType": "SC-QAM",
      "frequency": 658,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.6,
      "snr": 39,
    },
    {
      "channelId": "11",
      "channelType": "SC-QAM",
      "frequency": 666,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.4,
      "snr": 37,
    },
    {
      "channelId": "12",
      "channelType": "SC-QAM",
      "frequency": 674,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.1,
      "snr": 38,
    },
    {
      "channelId": "13",
      "channelType": "SC-QAM",
      "frequency": 682,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.7,
      "snr": 39,
    },
    {
      "channelId": "14",
      "channelType": "SC-QAM",
      "frequency": 690,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2,
      "snr": 39,
    },
    {
      "channelId": "15",
      "channelType": "SC-QAM",
      "frequency": 698,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -8,
      "snr": 34,
    },
    {
      "channelId": "16",
      "channelType": "SC-QAM",
      "frequency": 706,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7,
      "snr": 35,
    },
    {
      "channelId": "17",
      "channelType": "SC-QAM",
      "frequency": 714,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.8,
      "snr": 35,
    },
    {
      "channelId": "18",
      "channelType": "SC-QAM",
      "frequency": 722,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.1,
      "snr": 34,
    },
    {
      "channelId": "19",
      "channelType": "SC-QAM",
      "frequency": 730,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.1,
      "snr": 35,
    },
    {
      "channelId": "20",
      "channelType": "SC-QAM",
      "frequency": 738,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.1,
      "snr": 35,
    },
    {
      "channelId": "21",
      "channelType": "SC-QAM",
      "frequency": 746,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.6,
      "snr": 36,
    },
    {
      "channelId": "22",
      "channelType": "SC-QAM",
      "frequency": 754,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.2,
      "snr": 35,
    },
    {
      "channelId": "23",
      "channelType": "SC-QAM",
      "frequency": 762,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.9,
      "snr": 35,
    },
    {
      "channelId": "24",
      "channelType": "SC-QAM",
      "frequency": 770,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.2,
      "snr": 35,
    },
    {
      "channelId": "25",
      "channelType": "SC-QAM",
      "frequency": 778,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.9,
      "snr": 36,
    },
    {
      "channelId": "26",
      "channelType": "SC-QAM",
      "frequency": 786,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.7,
      "snr": 35,
    },
    {
      "channelId": "27",
      "channelType": "SC-QAM",
      "frequency": 794,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.9,
      "snr": 35,
    },
    {
      "channelId": "28",
      "channelType": "SC-QAM",
      "frequency": 802,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.4,
      "snr": 35,
    },
    {
      "channelId": "29",
      "channelType": "SC-QAM",
      "frequency": 810,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.5,
      "snr": 35,
    },
    {
      "channelId": "30",
      "channelType": "SC-QAM",
      "frequency": 818,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.3,
      "snr": 35,
    },
    {
      "channelId": "31",
      "channelType": "SC-QAM",
      "frequency": 826,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.8,
      "snr": 35,
    },
    {
      "channelId": "32",
      "channelType": "SC-QAM",
      "frequency": 834,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.9,
      "snr": 34,
    },
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 114,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -6.5,
      "snr": 34,
    },
  ],
  "downstreamOfdm": [
    {
      "channelId": "33",
      "channelType": "OFDM",
      "frequencyEnd": 324,
      "frequencyStart": 151,
      "lockStatus": "Locked",
      "modulation": "1024QAM",
      "powerLevel": -4.1,
      "snr": 39,
    },
  ],
  "time": "2021-04-11T13:38:36.163Z",
  "upstream": [
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 31,
      "lockStatus": "ACTIVE",
      "modulation": "64QAM",
      "powerLevel": 52.3,
      "snr": 0,
    },
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 51,
      "lockStatus": "ACTIVE",
      "modulation": "64QAM",
      "powerLevel": 52.3,
      "snr": 0,
    },
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 37,
      "lockStatus": "ACTIVE",
      "modulation": "32QAM",
      "powerLevel": 52.3,
      "snr": 0,
    },
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 45,
      "lockStatus": "ACTIVE",
      "modulation": "64QAM",
      "powerLevel": 52.3,
      "snr": 0,
    },
  ],
  "upstreamOfdma": [],
}
`;

exports[`Arris normalizeDocsisStatus should work with ofdm in download and in upload 1`] = `
{
  "downstream": [
    {
      "channelId": "5",
      "channelType": "SC-QAM",
      "frequency": 602,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.4,
      "snr": 38,
    },
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 114,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.7,
      "snr": 35,
    },
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 130,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5,
      "snr": 36,
    },
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 138,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -4.9,
      "snr": 36,
    },
    {
      "channelId": "31",
      "channelType": "SC-QAM",
      "frequency": 826,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.2,
      "snr": 35,
    },
    {
      "channelId": "32",
      "channelType": "SC-QAM",
      "frequency": 834,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.2,
      "snr": 34,
    },
    {
      "channelId": "30",
      "channelType": "SC-QAM",
      "frequency": 818,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.9,
      "snr": 35,
    },
    {
      "channelId": "29",
      "channelType": "SC-QAM",
      "frequency": 810,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.4,
      "snr": 35,
    },
    {
      "channelId": "7",
      "channelType": "SC-QAM",
      "frequency": 626,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.9,
      "snr": 38,
    },
    {
      "channelId": "6",
      "channelType": "SC-QAM",
      "frequency": 618,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -3.4,
      "snr": 39,
    },
    {
      "channelId": "26",
      "channelType": "SC-QAM",
      "frequency": 786,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.6,
      "snr": 34,
    },
    {
      "channelId": "27",
      "channelType": "SC-QAM",
      "frequency": 794,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.6,
      "snr": 35,
    },
    {
      "channelId": "28",
      "channelType": "SC-QAM",
      "frequency": 802,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6,
      "snr": 34,
    },
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 146,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -5.2,
      "snr": 36,
    },
    {
      "channelId": "25",
      "channelType": "SC-QAM",
      "frequency": 778,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.7,
      "snr": 35,
    },
    {
      "channelId": "9",
      "channelType": "SC-QAM",
      "frequency": 650,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.3,
      "snr": 39,
    },
    {
      "channelId": "10",
      "channelType": "SC-QAM",
      "frequency": 658,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.4,
      "snr": 39,
    },
    {
      "channelId": "11",
      "channelType": "SC-QAM",
      "frequency": 666,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.7,
      "snr": 38,
    },
    {
      "channelId": "22",
      "channelType": "SC-QAM",
      "frequency": 754,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.4,
      "snr": 35,
    },
    {
      "channelId": "23",
      "channelType": "SC-QAM",
      "frequency": 762,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.2,
      "snr": 35,
    },
    {
      "channelId": "24",
      "channelType": "SC-QAM",
      "frequency": 770,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.3,
      "snr": 35,
    },
    {
      "channelId": "8",
      "channelType": "SC-QAM",
      "frequency": 642,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2.4,
      "snr": 39,
    },
    {
      "channelId": "21",
      "channelType": "SC-QAM",
      "frequency": 746,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -5.4,
      "snr": 35,
    },
    {
      "channelId": "12",
      "channelType": "SC-QAM",
      "frequency": 674,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -2,
      "snr": 39,
    },
    {
      "channelId": "13",
      "channelType": "SC-QAM",
      "frequency": 682,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -1.9,
      "snr": 39,
    },
    {
      "channelId": "15",
      "channelType": "SC-QAM",
      "frequency": 698,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7.8,
      "snr": 34,
    },
    {
      "channelId": "16",
      "channelType": "SC-QAM",
      "frequency": 706,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.8,
      "snr": 35,
    },
    {
      "channelId": "18",
      "channelType": "SC-QAM",
      "frequency": 722,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -7,
      "snr": 35,
    },
    {
      "channelId": "14",
      "channelType": "SC-QAM",
      "frequency": 690,
      "lockStatus": "Locked",
      "modulation": "256QAM",
      "powerLevel": -1.5,
      "snr": 39,
    },
    {
      "channelId": "17",
      "channelType": "SC-QAM",
      "frequency": 714,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.8,
      "snr": 35,
    },
    {
      "channelId": "19",
      "channelType": "SC-QAM",
      "frequency": 730,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6.9,
      "snr": 34,
    },
    {
      "channelId": "20",
      "channelType": "SC-QAM",
      "frequency": 738,
      "lockStatus": "Locked",
      "modulation": "64QAM",
      "powerLevel": -6,
      "snr": 35,
    },
  ],
  "downstreamOfdm": [
    {
      "channelId": "33",
      "channelType": "OFDM",
      "frequencyEnd": 324,
      "frequencyStart": 151,
      "lockStatus": "Locked",
      "modulation": "1024QAM",
      "powerLevel": -3.8,
      "snr": 40,
    },
  ],
  "time": "2021-10-15T22:49:45.348Z",
  "upstream": [
    {
      "channelId": "3",
      "channelType": "SC-QAM",
      "frequency": 37,
      "lockStatus": "ACTIVE",
      "modulation": "32QAM",
      "powerLevel": 51,
      "snr": 0,
    },
    {
      "channelId": "1",
      "channelType": "SC-QAM",
      "frequency": 51,
      "lockStatus": "ACTIVE",
      "modulation": "32QAM",
      "powerLevel": 49,
      "snr": 0,
    },
    {
      "channelId": "4",
      "channelType": "SC-QAM",
      "frequency": 31,
      "lockStatus": "ACTIVE",
      "modulation": "32QAM",
      "powerLevel": 51,
      "snr": 0,
    },
    {
      "channelId": "2",
      "channelType": "SC-QAM",
      "frequency": 45,
      "lockStatus": "ACTIVE",
      "modulation": "32QAM",
      "powerLevel": 51,
      "snr": 0,
    },
  ],
  "upstreamOfdma": [
    {
      "channelId": "9",
      "channelType": "OFDMA",
      "frequencyEnd": 64.8,
      "frequencyStart": 29.8,
      "lockStatus": "SUCCESS",
      "modulation": "16_QAM",
      "powerLevel": 47,
      "snr": 0,
    },
  ],
}
`;
